go/store/nbs/archive_reader.go

```diff
@@ -0,0 +1,382 @@
+	sufRdr := io.NewSectionReader(reader, int64(suffixSpan.offset), int64(suffixSpan.length))
+	sufReader := bufio.NewReader(sufRdr)
+	suffixes := make([]suffix, footer.chunkCount)
+	for i := uint32(0); i < footer.chunkCount; i++ {
+		_, err := io.ReadFull(sufReader, suffixes[i][:])
+		if err != nil {
+			return archiveReader{}, err
+		}
+	}
+
+	return archiveReader{
+		reader:    reader,
+		prefixes:  prefixes,
+		byteSpans: byteSpans,
+		chunkRefs: chunks,
+		suffixes:  suffixes,
+		footer:    footer,
+	}, nil
+}
+
+func loadFooter(reader io.ReaderAt, fileSize uint64) (f footer, err error) {
+	section := io.NewSectionReader(reader, int64(fileSize-archiveFooterSize), int64(archiveFooterSize))
+
+	buf := make([]byte, archiveFooterSize)
+	_, err = io.ReadFull(section, buf)
+	if err != nil {
+		return
+	}
+
+	f.formatVersion = buf[afrVersionOffset]
+	f.fileSignature = string(buf[afrSigOffset:])
+	// Verify File Signature
+	if f.fileSignature != string(archiveFileSignature) {
+		err = ErrInvalidFileSignature
+		return
+	}
+	// Verify Format Version. Currently only one version is supported, but we'll need to be more flexible in the future.
+	if f.formatVersion != archiveFormatVersion {
+		err = ErrInvalidFormatVersion
+		return
+	}
+
+	f.indexSize = binary.BigEndian.Uint32(buf[afrIndexLenOffset : afrIndexChkSumOffset+uint32Size])
+	f.byteSpanCount = binary.BigEndian.Uint32(buf[afrByteSpanOffset : afrByteSpanOffset+uint32Size])
+	f.chunkCount = binary.BigEndian.Uint32(buf[afrChunkCountOffset : afrChunkCountOffset+uint32Size])
+	f.metadataSize = binary.BigEndian.Uint32(buf[afrMetaLenOffset : afrMetaLenOffset+uint32Size])
+	f.dataCheckSum = sha512Sum(buf[afrDataChkSumOffset : afrDataChkSumOffset+sha512.Size])
+	f.indexCheckSum = sha512Sum(buf[afrIndexChkSumOffset : afrIndexChkSumOffset+sha512.Size])
+	f.metaCheckSum = sha512Sum(buf[afrMetaChkSumOffset : afrMetaChkSumOffset+sha512.Size])
+	f.fileSize = fileSize
+
+	return
+}
+
+func (ai archiveReader) search(hash hash.Hash) (bool, int) {
+	prefix := hash.Prefix()
+	possibleMatch := prollyBinSearch(ai.prefixes, prefix)
+	targetSfx := hash.Suffix()
+
+	for i := 0; i < len(ai.suffixes); i++ {
+		idx := possibleMatch + i
+
+		if ai.prefixes[idx] != prefix {
+			return false, -1
+		}
+
+		if ai.suffixes[idx] == suffix(targetSfx) {
+			return true, idx
+		}
+	}
+	return false, -1
+}
+
+func (ai archiveReader) has(hash hash.Hash) bool {
+	found, _ := ai.search(hash)
+	return found
+}
+
+// get returns the decompressed data for the given hash. If the hash is not found, nil is returned (not an error)
+func (ai archiveReader) get(hash hash.Hash) ([]byte, error) {
+	dict, data, err := ai.getRaw(hash)
+	if err != nil || data == nil {
+		return nil, err
+	}
+
+	var result []byte
+	if dict == nil {
+		result, err = gozstd.Decompress(nil, data)
+	} else {
+		dDict, e2 := gozstd.NewDDict(dict)
+		if e2 != nil {
+			return nil, e2
+		}
+		result, err = gozstd.DecompressDict(nil, data, dDict)
```

Comment:
Do we want to `dict, err = gozstd.Decompress(nil, dict)` here? Are we going to have some compressed and some non-compressed dicts, or will all dicts be compressed or all dicts be uncompressed?

Reply:
All dicts will be compressed. Encoding which is and is not is not something we even have the ability to do currently.